Composer/packages/client/src/components/TestController/TestController.tsx

```diff
@@ -114,39 +119,75 @@ export const TestController: React.FC = () => {
     setCalloutVisible(true);
   }
 
-  async function handlePublishLuis(luisConfig) {
+  async function handlePublish(config) {
     setBotStatus(BotStatus.publishing);
     dismissDialog();
-    await setSettings(projectId, { ...settings, luis: luisConfig });
-    await publishLuis(luisConfig, projectId);
+    // save the settings change to store and persist to server
+    const newValue = config;
+    const subscriptionKey = newValue.subscriptionKey;
+    const qnaRegion = newValue.qnaRegion;
+    delete newValue.subscriptionKey;
+    delete newValue.qnaRegion;
+    await setSettings(projectId, {
+      ...settings,
+      luis: newValue,
+      qna: { ...settings.qna, subscriptionKey, region: qnaRegion, endpointKey: '' },
+    });
+    await build(newValue.authoringKey, subscriptionKey, qnaRegion, projectId);
   }
 
   async function handleLoadBot() {
     setBotStatus(BotStatus.reloading);
+    if (settings.qna && Object(settings.qna).subscriptionKey) {
+      await setQnASettings(projectId, Object(settings.qna).subscriptionKey);
+    }
     const sensitiveSettings = settingsStorage.get(projectId);
     await publishToTarget(projectId, DefaultPublishConfig, { comment: '' }, sensitiveSettings);
   }
 
-  function isLuisConfigComplete(config) {
+  function isConfigComplete(config) {
     let complete = true;
-    for (const key in LuisConfig) {
-      if (config?.[LuisConfig[key]] === '') {
-        complete = false;
-        break;
+    if (getReferredLuFiles(luFiles, dialogs).length > 0) {
+      for (const key in LuisConfig) {
+        if (config?.[LuisConfig[key]] === '') {
+          complete = false;
+          break;
+        }
+      }
+    }
+    if (getReferredQnaFiles(qnaFiles, dialogs).length > 0) {
+      for (const key in QnaConfig) {
+        if (config?.[QnaConfig[key]] === '') {
+          complete = false;
+          break;
+        }
       }
     }
     return complete;
   }
 
+  function isDialogDefaultRecognizer(dialogs) {
+    let isDefaultRecognizer = false;
+    dialogs.map((dialog) => {
+      if (typeof dialog.content.recognizer === 'string') {
+        isDefaultRecognizer = true;
+        return;
+      }
+    });
+    return isDefaultRecognizer;
+  }
+
   async function handleStart() {
     dismissCallout();
-    const config = settings.luis;
-
-    if (!isAbsHosted() && getReferredFiles(luFiles, dialogs).length > 0) {
-      if (botStatus === BotStatus.failed || botStatus === BotStatus.pending || !isLuisConfigComplete(config)) {
+    const config = Object.assign({}, settings.luis, {
+      subscriptionKey: Object(settings.qna).subscriptionKey,
+      qnaRegion: Object(settings.qna).qnaRegion,
+    });
+    if (!isAbsHosted() && isDialogDefaultRecognizer(dialogs)) {
```

Comment:
`isDialogDefaultRecognizer` is a confusing name. Rename it.

The hidden knowledge is: recognizer is a default one -> value is a string -> project contains qna / lu -> recognizer needs to be published before start.
No explicit relationship between 'default recognizer' and 'publish' logic.

Reply:
`needsPublish()`